src/algorithms/reco/ElectronReconstruction.cc

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Daniel Brandenburg
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+#include "Beam.h"
+#include "Boost.h"
+#include "ElectronReconstruction.h"
+
+// Event Model related classes
+#include "edm4hep/MCParticleCollection.h"
+#include "edm4eic/MCRecoParticleAssociationCollection.h"
+#include "edm4eic/ReconstructedParticleCollection.h"
+#include "edm4eic/MCRecoClusterParticleAssociationCollection.h"
+
+namespace eicrecon {
+
+  void ElectronReconstruction::init(std::shared_ptr<spdlog::logger> logger) {
+    m_log = logger;
+  }
+
+  std::unique_ptr<edm4eic::ReconstructedParticleCollection> ElectronReconstruction::execute(
+    const std::vector<const edm4hep::MCParticle *> &mcparts,
+    const std::vector<const edm4eic::ReconstructedParticle *> &rcparts,
+    const std::vector<const edm4eic::MCRecoParticleAssociation *> &rcassoc,
+    const std::vector<const edm4eic::MCRecoClusterParticleAssociationCollection*> &in_clu_assoc
+    ) {
+
+        // Step 1. Loop through MCParticle - cluster associations
+        // Step 2. Get Reco particle for the Mc Particle matched to cluster
+        // Step 3. Apply E/p cut using Reco cluster Energy and Reco Particle momentum
+
+        // Some obvious improvements:
+        // - E/p cut from real study optimized for electron finding and hadron rejection
+        // - use of any HCAL info?
+        // - check for duplicates?
+
+        // output container
+        auto out_electrons = std::make_unique<edm4eic::ReconstructedParticleCollection>();
+
+        for ( auto col : in_clu_assoc ){ // loop on cluster association collections
+          for ( size_t i = 0; i < col->size(); i++ ){ // loop on MCRecoClusterParticleAssociation in this particular collection
+            const auto& clu_assoc = col->at(i);
+            auto sim = clu_assoc.getSim(); // McParticle
+            auto clu = clu_assoc.getRec(); // RecoCluster
+
+            m_log->trace( "SimId={}, CluId={}", clu_assoc.getSimID(), clu_assoc.getRecID() );
+            m_log->trace( "MCParticle: Energy={} GeV, p={} GeV, E/p = {} for PDG: {}", clu.getEnergy(), edm4eic::magnitude(sim.getMomentum()), clu.getEnergy() / edm4eic::magnitude(sim.getMomentum()), sim.getPDG() );
+
+
+            // Find the Reconstructed particle associated to the MC Particle that is matched with this reco cluster
+            // i.e. take (MC Particle <-> RC Cluster) + ( MC Particle <-> RC Particle ) = ( RC Particle <-> RC Cluster )
+            auto reco_part_assoc = rcassoc.begin();
+            bool found_reco_part = false;
```

Comment:
Prefer to loop without sentinel and use equality with end as marker.

Reply:
done with bf659cb